File: clients/create/templates/android-minsdk24/feature/login/src/main/kotlin/io/github/ydonghao/{{appPackageSegment}}/feature/login/LoginViewModel.kt
package io.github.ydonghao.{{appPackageSegment}}.feature.login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * 每屏一个 ViewModel = 唯一状态容器（contract/clients/android.md 二-2）：
 * UiState 单一可信状态源 + 一次性事件通道；UI 只发事件不改状态（UDF）。
 */
@HiltViewModel
class LoginViewModel @Inject constructor(
    private val repository: LoginRepository
) : ViewModel() {

    data class UiState(
        val username: String = "",
        val password: String = "",
        val loading: Boolean = false,
        val errorMessage: String? = null
    )

    sealed interface LoginEvent {
        data object LoggedIn : LoginEvent
    }

    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state.asStateFlow()

    private val _events = Channel<LoginEvent>(Channel.BUFFERED)
    val events: Flow<LoginEvent> = _events.receiveAsFlow()

    fun onUsernameChange(value: String) {
        _state.update { it.copy(username = value, errorMessage = null) }
    }

    fun onPasswordChange(value: String) {
        _state.update { it.copy(password = value, errorMessage = null) }
    }

    fun submit() {
        if (_state.value.loading) {
            return
        }
        viewModelScope.launch {
            _state.update { it.copy(loading = true, errorMessage = null) }
            when (val result = repository.login(_state.value.username, _state.value.password)) {
                is LoginResult.Success -> {
                    _state.update { it.copy(loading = false) }
                    _events.send(LoginEvent.LoggedIn)
                }
                is LoginResult.Failure -> {
                    _state.update { it.copy(loading = false, errorMessage = result.message) }
                }
            }
        }
    }
}

File: clients/create/templates/ios17/Packages/Features/Login/Sources/Login/LoginViewModel.swift
import Foundation
import Observation

/**
 * MVVM + @Observable（M6 拍板；contract/clients/ios.md 二-2 ios17 档正式口径）：
 * ViewModel 持全部可变状态（禁 import SwiftUI，可测性前提）；View 只发意图。
 * 登录成功是状态迁移（loggedIn），导航由 View 消费该迁移一次性触发。
 */
@MainActor
public final class LoginViewModel: Observable {
    public struct State: Sendable, Equatable {
        public var username = ""
        public var password = ""
        public var loading = false
        public var errorMessage: String?
        public var loggedIn = false

        public init() {}
    }

    // setter 开放给 SwiftUI Binding 写路径（@Bindable 投影）；写语义仍收敛于 VM 的意图方法，
    // View 不直接改 loading/loggedIn 等流程字段
    public var state = State()

    private let repository: any LoginRepository

    public init(repository: any LoginRepository = MockLoginRepository()) {
        self.repository = repository
    }

    public func onUsernameChange(_ value: String) {
        state.username = value
        state.errorMessage = nil
    }

    public func onPasswordChange(_ value: String) {
        state.password = value
        state.errorMessage = nil
    }

    public func submit() async {
        guard !state.loading else { return }
        state.loading = true
        state.errorMessage = nil
        let result = await repository.login(username: state.username, password: state.password)
        state.loading = false
        switch result {
        case .success:
            state.loggedIn = true
        case .failure(let message):
            state.errorMessage = message
        }
    }
}
